File: implementations/rust/src/hash.rs
//! BLAKE3-256 + multihash wrapping (SPEC-1 §4, ERRATA D7). Mirrors ../ts/src/hash.ts.

const BLAKE3_MULTICODEC: u8 = 0x1e;

/// id = multihash(BLAKE3-256(data)) as lowercase hex.
/// The multicodec (0x1e) and length (32 = 0x20) each fit a single-byte varint.
pub fn content_address(data: &[u8]) -> String {
    let digest = blake3::hash(data);
    let bytes = digest.as_bytes(); // &[u8; 32]
    let mut mh = Vec::with_capacity(2 + bytes.len());
    mh.push(BLAKE3_MULTICODEC);
    mh.push(bytes.len() as u8);
    mh.extend_from_slice(bytes);
    hex::encode(mh)
}

File: implementations/ts/src/delta/hash.ts
import { blake3 } from "@noble/hashes/blake3";
import { bytesToHex } from "@noble/hashes/utils";

// BLAKE3 multicodec code (https://github.com/multiformats/multicodec). Both it and the 32-byte
// length fit in a single-byte varint, so the multihash prefix is simply [0x1e, 0x20].
const BLAKE3_MULTICODEC = 0x1e;
const DIGEST_LEN = 32;

// id = multihash(BLAKE3-256(data)) as lowercase hex (SPEC-1 §4, ERRATA D7).
export function contentAddress(data: Uint8Array): string {
  const digest = blake3(data, { dkLen: DIGEST_LEN });
  const mh = new Uint8Array(2 + digest.length);
  mh[0] = BLAKE3_MULTICODEC;
  mh[1] = digest.length;
  mh.set(digest, 2);
  return bytesToHex(mh);
}

export { bytesToHex };
